File: java/rocksjni/ingest_external_file_options.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
//
// This file implements the "bridge" between Java and C++ for
// ROCKSDB_NAMESPACE::FilterPolicy.

#include <jni.h>

#include "include/org_rocksdb_IngestExternalFileOptions.h"
#include "rocksdb/options.h"
#include "rocksjni/cplusplus_to_java_convert.h"

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    newIngestExternalFileOptions
 * Signature: ()J
 */
jlong Java_org_rocksdb_IngestExternalFileOptions_newIngestExternalFileOptions__(
    JNIEnv*, jclass) {
  auto* options = new ROCKSDB_NAMESPACE::IngestExternalFileOptions();
  return GET_CPLUSPLUS_POINTER(options);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    newIngestExternalFileOptions
 * Signature: (ZZZZ)J
 */
jlong Java_org_rocksdb_IngestExternalFileOptions_newIngestExternalFileOptions__ZZZZ(
    JNIEnv*, jclass, jboolean jmove_files, jboolean jsnapshot_consistency,
    jboolean jallow_global_seqno, jboolean jallow_blocking_flush) {
  auto* options = new ROCKSDB_NAMESPACE::IngestExternalFileOptions();
  options->move_files = static_cast<bool>(jmove_files);
  options->snapshot_consistency = static_cast<bool>(jsnapshot_consistency);
  options->allow_global_seqno = static_cast<bool>(jallow_global_seqno);
  options->allow_blocking_flush = static_cast<bool>(jallow_blocking_flush);
  return GET_CPLUSPLUS_POINTER(options);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    moveFiles
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_moveFiles(JNIEnv*, jclass,
                                                              jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return static_cast<jboolean>(options->move_files);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setMoveFiles
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setMoveFiles(
    JNIEnv*, jclass, jlong jhandle, jboolean jmove_files) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->move_files = static_cast<bool>(jmove_files);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    snapshotConsistency
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_snapshotConsistency(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return static_cast<jboolean>(options->snapshot_consistency);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setSnapshotConsistency
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setSnapshotConsistency(
    JNIEnv*, jclass, jlong jhandle, jboolean jsnapshot_consistency) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->snapshot_consistency = static_cast<bool>(jsnapshot_consistency);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    allowGlobalSeqNo
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_allowGlobalSeqNo(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return static_cast<jboolean>(options->allow_global_seqno);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setAllowGlobalSeqNo
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setAllowGlobalSeqNo(
    JNIEnv*, jclass, jlong jhandle, jboolean jallow_global_seqno) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->allow_global_seqno = static_cast<bool>(jallow_global_seqno);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    allowBlockingFlush
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_allowBlockingFlush(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return static_cast<jboolean>(options->allow_blocking_flush);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setAllowBlockingFlush
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setAllowBlockingFlush(
    JNIEnv*, jclass, jlong jhandle, jboolean jallow_blocking_flush) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->allow_blocking_flush = static_cast<bool>(jallow_blocking_flush);
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    ingestBehind
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_ingestBehind(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->ingest_behind == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setIngestBehind
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setIngestBehind(
    JNIEnv*, jclass, jlong jhandle, jboolean jingest_behind) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->ingest_behind = jingest_behind == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    writeGlobalSeqno
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_writeGlobalSeqno(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->write_global_seqno == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setWriteGlobalSeqno
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setWriteGlobalSeqno(
    JNIEnv*, jclass, jlong jhandle, jboolean jwrite_global_seqno) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->write_global_seqno = jwrite_global_seqno == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    verifyChecksumsBeforeIngest
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_verifyChecksumsBeforeIngest(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->verify_checksums_before_ingest ? JNI_TRUE : JNI_FALSE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setVerifyChecksumsBeforeIngest
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setVerifyChecksumsBeforeIngest(
    JNIEnv*, jclass, jlong jhandle, jboolean jverify_checksums_before_ingest) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->verify_checksums_before_ingest =
      jverify_checksums_before_ingest == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    verifyChecksumsReadaheadSize
 * Signature: (J)Z
 */
jlong Java_org_rocksdb_IngestExternalFileOptions_verifyChecksumsReadaheadSize(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->verify_checksums_readahead_size;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setVerifyChecksumsReadaheadSize
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setVerifyChecksumsReadaheadSize(
    JNIEnv*, jclass, jlong jhandle, jlong jverify_checksums_readahead_size) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->verify_checksums_readahead_size = jverify_checksums_readahead_size;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    verifyFileChecksum
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_verifyFileChecksum(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->verify_file_checksum ? JNI_TRUE : JNI_FALSE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setVerifyFileChecksum
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setVerifyFileChecksum(
    JNIEnv*, jclass, jlong jhandle, jboolean jverify_file_checksum) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->verify_file_checksum = jverify_file_checksum == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    failIfNotLastLevel
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_failIfNotLastLevel(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->fail_if_not_bottommost_level ? JNI_TRUE : JNI_FALSE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setFailIfNotLastLevel
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setFailIfNotLastLevel(
    JNIEnv*, jclass, jlong jhandle, jboolean jfail_if_not_bottommost_level) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->fail_if_not_bottommost_level =
      jfail_if_not_bottommost_level == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    linkFiles
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_IngestExternalFileOptions_linkFiles(JNIEnv*, jclass,
                                                              jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  return options->link_files ? JNI_TRUE : JNI_FALSE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    setLinkFiles
 * Signature: (JZ)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_setLinkFiles(
    JNIEnv*, jclass, jlong jhandle, jboolean jlink_files) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  options->link_files = jlink_files == JNI_TRUE;
}

/*
 * Class:     org_rocksdb_IngestExternalFileOptions
 * Method:    disposeInternal
 * Signature: (J)V
 */
void Java_org_rocksdb_IngestExternalFileOptions_disposeInternalJni(
    JNIEnv*, jclass, jlong jhandle) {
  auto* options =
      reinterpret_cast<ROCKSDB_NAMESPACE::IngestExternalFileOptions*>(jhandle);
  delete options;
}

File: java/src/main/java/org/rocksdb/IngestExternalFileOptions.java
package org.rocksdb;
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

import java.util.List;

/**
 * IngestExternalFileOptions is used by
 * {@link RocksDB#ingestExternalFile(ColumnFamilyHandle, List, IngestExternalFileOptions)}.
 */
public class IngestExternalFileOptions extends RocksObject {

  public IngestExternalFileOptions() {
    super(newIngestExternalFileOptions());
  }

  /**
   * @param moveFiles {@link #setMoveFiles(boolean)}
   * @param snapshotConsistency {@link #setSnapshotConsistency(boolean)}
   * @param allowGlobalSeqNo {@link #setAllowGlobalSeqNo(boolean)}
   * @param allowBlockingFlush {@link #setAllowBlockingFlush(boolean)}
   */
  public IngestExternalFileOptions(final boolean moveFiles,
      final boolean snapshotConsistency, final boolean allowGlobalSeqNo,
      final boolean allowBlockingFlush) {
    super(newIngestExternalFileOptions(moveFiles, snapshotConsistency,
        allowGlobalSeqNo, allowBlockingFlush));
  }

  /**
   * Can be set to true to move the files instead of copying them.
   *
   * @return true if files will be moved
   */
  public boolean moveFiles() {
    return moveFiles(nativeHandle_);
  }

  /**
   * Can be set to true to move the files instead of copying them.
   *
   * @param moveFiles true if files should be moved instead of copied
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setMoveFiles(final boolean moveFiles) {
    setMoveFiles(nativeHandle_, moveFiles);
    return this;
  }

  /**
   * If set to false, an ingested file keys could appear in existing snapshots
   * that where created before the file was ingested.
   *
   * @return true if snapshot consistency is assured
   */
  public boolean snapshotConsistency() {
    return snapshotConsistency(nativeHandle_);
  }

  /**
   * If set to false, an ingested file keys could appear in existing snapshots
   * that where created before the file was ingested.
   *
   * @param snapshotConsistency true if snapshot consistency is required
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setSnapshotConsistency(
      final boolean snapshotConsistency) {
    setSnapshotConsistency(nativeHandle_, snapshotConsistency);
    return this;
  }

  /**
   * If set to false, {@link RocksDB#ingestExternalFile(ColumnFamilyHandle, List, IngestExternalFileOptions)}
   * will fail if the file key range overlaps with existing keys or tombstones in the DB.
   *
   * @return true if global seq numbers are assured
   */
  public boolean allowGlobalSeqNo() {
    return allowGlobalSeqNo(nativeHandle_);
  }

  /**
   * If set to false, {@link RocksDB#ingestExternalFile(ColumnFamilyHandle, List, IngestExternalFileOptions)}
   * will fail if the file key range overlaps with existing keys or tombstones in the DB.
   *
   * @param allowGlobalSeqNo true if global seq numbers are required
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setAllowGlobalSeqNo(
      final boolean allowGlobalSeqNo) {
    setAllowGlobalSeqNo(nativeHandle_, allowGlobalSeqNo);
    return this;
  }

  /**
   * If set to false and the file key range overlaps with the memtable key range
   * (memtable flush required), IngestExternalFile will fail.
   *
   * @return true if blocking flushes may occur
   */
  public boolean allowBlockingFlush() {
    return allowBlockingFlush(nativeHandle_);
  }

  /**
   * If set to false and the file key range overlaps with the memtable key range
   * (memtable flush required), IngestExternalFile will fail.
   *
   * @param allowBlockingFlush true if blocking flushes are allowed
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setAllowBlockingFlush(
      final boolean allowBlockingFlush) {
    setAllowBlockingFlush(nativeHandle_, allowBlockingFlush);
    return this;
  }

  /**
   * Returns true if duplicate keys in the file being ingested are
   * to be skipped rather than overwriting existing data under that key.
   *
   * @return true if duplicate keys in the file being ingested are to be
   *     skipped, false otherwise.
   */
  public boolean ingestBehind() {
    return ingestBehind(nativeHandle_);
  }

  /**
   * Set to true if you would like duplicate keys in the file being ingested
   * to be skipped rather than overwriting existing data under that key.
   * <p>
   * Usecase: back-fill of some historical data in the database without
   * over-writing existing newer version of data.
   * <p>
   * This option could only be used if the DB has been running
   * with DBOptions#allowIngestBehind() == true since the dawn of time.
   * <p>
   * All files will be ingested at the bottommost level with seqno=0.
   * <p>
   * Default: false
   *
   * @param ingestBehind true if you would like duplicate keys in the file being
   *     ingested to be skipped.
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setIngestBehind(final boolean ingestBehind) {
    setIngestBehind(nativeHandle_, ingestBehind);
    return this;
  }

  /**
   * Returns true write if the global_seqno is written to a given offset
   * in the external SST file for backward compatibility.
   * <p>
   * See {@link #setWriteGlobalSeqno(boolean)}.
   *
   * @return true if the global_seqno is written to a given offset,
   *     false otherwise.
   */
  public boolean writeGlobalSeqno() {
    return writeGlobalSeqno(nativeHandle_);
  }

  /**
   * Set to true if you would like to write the global_seqno to a given offset
   * in the external SST file for backward compatibility.
   * <p>
   * Older versions of RocksDB write the global_seqno to a given offset within
   * the ingested SST files, and new versions of RocksDB do not.
   * <p>
   * If you ingest an external SST using new version of RocksDB and would like
   * to be able to downgrade to an older version of RocksDB, you should set
   * {@link #writeGlobalSeqno()} to true.
   * <p>
   * If your service is just starting to use the new RocksDB, we recommend that
   * you set this option to false, which brings two benefits:
   *    1. No extra random write for global_seqno during ingestion.
   *    2. Without writing external SST file, it's possible to do checksum.
   * <p>
   * We have a plan to set this option to false by default in the future.
   * <p>
   * Default: true
   *
   * @param writeGlobalSeqno true to write the gloal_seqno to a given offset,
   *     false otherwise
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setWriteGlobalSeqno(
      final boolean writeGlobalSeqno) {
    setWriteGlobalSeqno(nativeHandle_, writeGlobalSeqno);
    return this;
  }

  /**
   * True iff the option to verify the checksums of each block of the
   * external SST file before ingestion has been set.
   *
   * @return true iff the option to verify the checksums of each block of the
   * external SST file before ingestion has been set.
   */
  public boolean verifyChecksumsBeforeIngest() {
    return verifyChecksumsBeforeIngest(nativeHandle_);
  }

  /**
   * Set to true if you would like to verify the checksums of each block of the
   * external SST file before ingestion.
   * <p>
   * Warning: setting this to true causes slowdown in file ingestion because
   * the external SST file has to be read.
   * <p>
   * Default: false
   *
   * @param verifyChecksumsBeforeIngest true if you would like to verify the checksums of each block
   *     of the
   * external SST file before ingestion.
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setVerifyChecksumsBeforeIngest(
      final boolean verifyChecksumsBeforeIngest) {
    setVerifyChecksumsBeforeIngest(nativeHandle_, verifyChecksumsBeforeIngest);
    return this;
  }

  /**
   * When verify_checksums_before_ingest = true, RocksDB uses default
   * readahead setting to scan the file while verifying checksums before
   * ingestion
   * <p>
   * Users can override the default value using this option.
   * Using a large readahead size (&gt; 2MB) can typically improve the performance
   * of forward iteration on spinning disks.
   *
   * @return the current value of readahead size (0 if it has not been set)
   */
  public long verifyChecksumsReadaheadSize() {
    return verifyChecksumsReadaheadSize(nativeHandle_);
  }

  /**
   * When verify_checksums_before_ingest = true, RocksDB uses default
   * readahead setting to scan the file while verifying checksums before
   * ingestion
   * <p>
   * Users can override the default value using this option.
   * Using a large readahead size (&gt; 2MB) can typically improve the performance
   * of forward iteration on spinning disks.
   *
   * @param verifyChecksumsReadaheadSize the value of readahead size to set
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setVerifyChecksumsReadaheadSize(
      final long verifyChecksumsReadaheadSize) {
    setVerifyChecksumsReadaheadSize(nativeHandle_, verifyChecksumsReadaheadSize);
    return this;
  }

  /**
   * Set to TRUE if user wants to verify the sst file checksum of ingested
   * files. The DB checksum function will generate the checksum of each
   * ingested file (if file_checksum_gen_factory is set) and compare the
   * checksum function name and checksum with the ingested checksum information.
   *
   * @return true iff the option to verify the sst file checksum of ingested files is set
   */
  public boolean verifyFileChecksum() {
    return verifyFileChecksum(nativeHandle_);
  }

  /**
   * Set to TRUE if user wants to verify the sst file checksum of ingested
   * files. The DB checksum function will generate the checksum of each
   * ingested file (if file_checksum_gen_factory is set) and compare the
   * checksum function name and checksum with the ingested checksum information.
   * <p>
   * If this option is set to True: 1) if DB does not enable checksum
   * (file_checksum_gen_factory == nullptr), the ingested checksum information
   * will be ignored; 2) If DB enable the checksum function, we calculate the
   * sst file checksum after the file is moved or copied and compare the
   * checksum and checksum name. If checksum or checksum function name does
   * not match, ingestion will be failed. If the verification is successful,
   * checksum and checksum function name will be stored in Manifest.
   * If this option is set to FALSE, 1) if DB does not enable checksum,
   * the ingested checksum information will be ignored; 2) if DB enable the
   * checksum, we only verify the ingested checksum function name and we
   * trust the ingested checksum. If the checksum function name matches, we
   * store the checksum in Manifest. DB does not calculate the checksum during
   * ingestion. However, if no checksum information is provided with the
   * ingested files, DB will generate the checksum and store in the Manifest.
   *
   * @param verifyFileChecksum true iff user wants to verify the sst file checksum of ingested files
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setVerifyFileChecksum(final boolean verifyFileChecksum) {
    setVerifyFileChecksum(nativeHandle_, verifyFileChecksum);
    return this;
  }

  /**
   * Set to TRUE if user wants file to be ingested to the last level. An
   * error of Status::TryAgain() will be returned if a file cannot fit in the
   * last level when calling {@link RocksDB#ingestExternalFile(List, IngestExternalFileOptions)}
   * or {@link RocksDB#ingestExternalFile(ColumnFamilyHandle, List, IngestExternalFileOptions)}.
   * The user should clear
   * the last level in the overlapping range before re-attempt.
   * <p>
   * @return true iff failIfNotLastLevel option has been set
   */
  public boolean failIfNotLastLevel() {
    return failIfNotLastLevel(nativeHandle_);
  }

  /**
   * Set to TRUE if user wants file to be ingested to the last level. An
   * error of Status::TryAgain() will be returned if a file cannot fit in the
   * last level when calling
   * DB::IngestExternalFile()/DB::IngestExternalFiles(). The user should clear
   * the last level in the overlapping range before re-attempt.
   * <p>
   * ingestBehind takes precedence over failIfNotLastLevel.
   * <p>
   * This method is named for "last" instead of "bottommost" (in the C++ API)
   * as that API notes that "bottommost" is obsolete/confusing terminology to refer to last level
   *
   * @param failIfNotLastLevel true iff user wants file to be ingested to the last level
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setFailIfNotLastLevel(final boolean failIfNotLastLevel) {
    setFailIfNotLastLevel(nativeHandle_, failIfNotLastLevel);
    return this;
  }

  /**
   * True if the files will be linked instead of copying them.
   * Same as moveFiles except that input files will NOT be unlinked.
   * Only one of `moveFiles` and `linkFiles` can be set at the same time.
   *
   * @return true if files will be moved
   */
  public boolean linkFiles() {
    return linkFiles(nativeHandle_);
  }

  /**
   * Can be set to true to link the files instead of copying them.
   * Same as moveFiles except that input files will NOT be unlinked.
   * Only one of `moveFiles` and `linkFiles` can be set at the same time.
   *
   * @param linkFiles true if files should be linked instead of copied
   *
   * @return the reference to the current IngestExternalFileOptions.
   */
  public IngestExternalFileOptions setLinkFiles(final boolean linkFiles) {
    setLinkFiles(nativeHandle_, linkFiles);
    return this;
  }

  private static native long newIngestExternalFileOptions();
  private static native long newIngestExternalFileOptions(final boolean moveFiles,
      final boolean snapshotConsistency, final boolean allowGlobalSeqNo,
      final boolean allowBlockingFlush);
  @Override
  protected final void disposeInternal(final long handle) {
    disposeInternalJni(handle);
  }

  private static native void disposeInternalJni(final long handle);

  private static native boolean moveFiles(final long handle);
  private static native void setMoveFiles(final long handle, final boolean move_files);
  private static native boolean snapshotConsistency(final long handle);
  private static native void setSnapshotConsistency(
      final long handle, final boolean snapshotConsistency);
  private static native boolean allowGlobalSeqNo(final long handle);
  private static native void setAllowGlobalSeqNo(final long handle, final boolean allowGloablSeqNo);
  private static native boolean allowBlockingFlush(final long handle);
  private static native void setAllowBlockingFlush(
      final long handle, final boolean allowBlockingFlush);
  private static native boolean ingestBehind(final long handle);
  private static native void setIngestBehind(final long handle, final boolean ingestBehind);
  private static native boolean writeGlobalSeqno(final long handle);
  private static native void setWriteGlobalSeqno(final long handle, final boolean writeGlobalSeqNo);
  private static native boolean verifyChecksumsBeforeIngest(final long handle);
  private static native void setVerifyChecksumsBeforeIngest(
      final long handle, final boolean verifyChecksumsBeforeIngest);
  private static native long verifyChecksumsReadaheadSize(final long handle);
  private static native void setVerifyChecksumsReadaheadSize(
      final long handle, final long verifyChecksumsReadaheadSize);
  private static native boolean verifyFileChecksum(final long handle);
  private static native void setVerifyFileChecksum(
      final long handle, final boolean verifyFileChecksum);
  private static native boolean failIfNotLastLevel(final long handle);
  private static native void setFailIfNotLastLevel(
      final long handle, final boolean failIfNotLastLevel);

  private static native boolean linkFiles(final long handle);
  private static native void setLinkFiles(final long handle, final boolean linkFiles);
}
